src/main/java/org/mule/module/apikit/odata/processor/ODataApikitProcessor.java

```diff
@@ -134,61 +130,68 @@ private String encode(String id) throws ODataInvalidUriException {
   }
 
   @Override
-  public ODataPayload process(CoreEvent event, AbstractRouterInterface router, List<Format> formats)
-      throws Exception {
-    String oDataURL = getCompleteUrl(CoreEventUtils.getHttpRequestAttributes(event));
+  public Mono<ODataPayload<?>> process(CoreEvent event, AbstractRouterInterface router,
+      List<Format> formats) {
+    String oDataCompleteURL = getCompleteUrl(CoreEventUtils.getHttpRequestAttributes(event));
 
     // truncate the URL at the entity
-    oDataURL = ODataUriHelper.getOdataUrl(oDataURL);
+    String oDataURL = ODataUriHelper.getOdataUrl(oDataCompleteURL);
 
     // invoke flow and validate response
-    ODataPayload<List<Entry>> oDataPayload = processEntityRequest(event, router, formats);
-    final List<Entry> entries = oDataPayload.getValue();
-
-    if (isEntityCount()) {
-      if (formats.contains(Format.Plain) || formats.contains(Format.Default)) {
-        String count = String.valueOf(entries.size());
-        return new ODataPayload<String>(oDataPayload.getMuleEvent(), count,
-            oDataPayload.getStatus());
+    return processEntityRequest(event, router, formats).flatMap((oDataPayload) -> {
+      final List<Entry> entries = oDataPayload.getValue();
+
+      if (isEntityCount()) {
```

Comment:
I would go with the negative case first to avoid nested blocks here. This seems to be the same:
```
if (!isEntityCount()) {
  oDataPayload.setFormatter(new ODataApiKitFormatter(getMetadataManager(), entity, oDataURL,
            entries, oDataPayload.getInlineCount(getInlineCountParam(event))));
  return Mono.just(oDataPayload);
}

if (formats.contains(Format.Plain) || formats.contains(Format.Default)) {
  String count = String.valueOf(entries.size());
  return Mono.just(
    new ODataPayload<>(oDataPayload.getMuleEvent(), count, oDataPayload.getStatus()));
} else {
  return Mono
    .error(new ODataUnsupportedMediaTypeException("Unsupported media type requested."));
}
```

Reply:
That's a good idea!